packages/@lwc/ssr-compiler/src/compile-template/shared.ts

```diff
@@ -25,41 +27,108 @@ import type {
     ObjectExpression as EsObjectExpression,
     Property as EsProperty,
     Statement as EsStatement,
+    IfStatement as EsIfStatement,
+    YieldExpression as EsYieldExpression,
+    ExpressionStatement as EsExpressionStatement,
 } from 'estree';
 import type {
     ComplexExpression as IrComplexExpression,
     Expression as IrExpression,
     Literal as IrLiteral,
 } from '@lwc/template-compiler';
 
+const bYieldTernary =
+    esTemplateWithYield`yield ${is.expression} ? ${is.expression} : ${is.expression}`<EsExpressionStatement>;
+
+interface OptimizableYield extends EsExpressionStatement {
+    expression: EsYieldExpression & { delegate: false };
+}
+
+const bOptimizedYield = esTemplateWithYield`yield ${is.expression};`<OptimizableYield>;
+
+function isOptimizableYield(stmt: EsStatement | undefined): stmt is OptimizableYield {
+    return (
+        is.expressionStatement(stmt) &&
+        is.yieldExpression(stmt.expression) &&
+        stmt.expression.delegate === false
+    );
+}
+
+/** Returns null if the statement cannot be optimized. */
+function optimizeSingleStatement(stmt: EsStatement): OptimizableYield | null {
+    if (is.blockStatement(stmt)) {
+        // `if (cond) { ... }` => optimize inner yields and see if we can condense
+        const optimizedBlock = optimizeAdjacentYieldStmts(stmt.body);
+        // More than one statement cannot be optimized into a single yield
+        if (optimizedBlock.length !== 1) return null;
```

Comment:
Out of curiousity, when would the `optimizedBlock.length` be more than 1? 

Is this when there are nested scopes, eg. nested if statements?

Reply:
It would be more than 1 for any regular `if` block. This is searching the specific case of `if (condition) { yield something }` and ignoring everything else.